### PR TITLE
refactor RasaModel and nlu

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- reach level of abstraction for `RasaModel`
- utilize same methods in train and predict graphs
- part of https://github.com/RasaHQ/research/issues/62

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 1]
```diff
@@ -1007,11 +1028,10 @@ def _prepare_entity_recognition_layers(self) -> None:
             average="micro",
         )
 
-    def set_training_phase(self, training: bool) -> None:
-        if training:
-            self.training = tf.ones((), tf.bool)
-        else:
-            self.training = tf.zeros((), tf.bool)
+    @staticmethod
+    def _get_mask_and_lengths(mask):
```
Comment (tabergma) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: please add types

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 2]
```diff
@@ -1007,11 +1028,10 @@ def _prepare_entity_recognition_layers(self) -> None:
...
+        sequence_lengths = tf.cast(tf.reduce_sum(mask[:, :, 0], 1), tf.int32)
+        return mask, sequence_lengths
```
Comment (tabergma) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: `mask` is not modified, why does the function return it?

File: rasa/utils/tensorflow/tf_models.py [hunk 3]
```diff
@@ -106,96 +113,117 @@ def fit(
...
-                tf_evaluation_dataset_function = tf.function(
-                    func=evaluation_dataset_function
-                )
-                tf_evaluation_on_batch_function = tf.function(
-                    self.evaluate_on_batch,
-                    input_signature=[
-                        tf_evaluation_dataset_function(tf_batch_size).element_spec
-                    ],
+
+            def evaluation_dataset_function(_batch_size):
```
Comment (tabergma) [inline rasa/utils/tensorflow/tf_models.py]: we could also add types here

File: rasa/nlu/classifiers/embedding_intent_classifier.py (hunk 2 above)
Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: because I call it 2 times, once for predict, once for train, so I thought it is less error prone to reduce the call to 1 line

Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: we could simply save mask into a local variable before calling the function

Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: what do you mean?

Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: instead of 
```
        mask_text, sequence_lengths = self._get_mask_and_lengths(
            tf_batch_data["text_mask"][0]
        )
```
use
```
        mask_text = tf_batch_data["text_mask"][0]
        sequence_lengths = self._get_mask_and_lengths(
            mask_text
        )
```

File: rasa/utils/tensorflow/tf_models.py [hunk 4]
```diff
@@ -106,96 +117,169 @@ def fit(
...
+        ) -> tf.data.Dataset:
+            return predict_data.as_tf_dataset(_batch_size, "sequence", shuffle=False)
+
+        _, self._predict_function = self._get_tf_functions(
+            predict_dataset_function, self.batch_predict, eager, "prediction"
+        )
+
+    def predict(self, predict_data: RasaModelData) -> Dict[Text, tf.Tensor]:
+        if self._predict_function is None:
+            logger.debug("There is no tensorflow prediction graph")
```
Comment (tabergma) [inline rasa/utils/tensorflow/tf_models.py]: The debug method might be misleading. Maybe something like `No tensorflow prediction graph present. Building it now.`

File: rasa/utils/tensorflow/tf_models.py [hunk 5]
```diff
@@ -106,96 +117,169 @@ def fit(
...
+
+    @staticmethod
+    def _get_tf_functions(
+        dataset_function: Callable,
+        call_model_function: Callable,
         eager: bool,
-        model_data: RasaModelData,
-        batch_strategy: Text,
-        tf_batch_size: tf.Tensor,
+        method: Text,
```
Comment (tabergma) [inline rasa/utils/tensorflow/tf_models.py]: rename `method` to `phase` or `state`?

File: rasa/utils/tensorflow/tf_models.py (hunk 4 above)
Comment (tabergma) [reply rasa/utils/tensorflow/tf_models.py]: Ah, not needed. Will be logged in `_get_tf_functions`. But can you please add a `.` at the end of the log statements. Also for the others further down.
